src/main/scala/com/advancedtelematic/campaigner/http/CampaignResource.scala

```diff
@@ -37,6 +40,25 @@ class CampaignResource(extractAuth: Directive1[AuthedNamespaceScope],
     } yield campaignId
   }
 
+  /**
+    * Create and immediately launch a retry-campaign for all the devices that were processed by `mainCampaign` and
+    * failed with a code `failureCode`. We assume there is at least one failed device. Hence, if there were not any
+    * such devices in `mainCampaign`, throw an error.
+    */
+  def retryFailedDevices(ns: Namespace, mainCampaign: Campaign, request: RetryFailedDevices): Future[CampaignId] =
+    campaigns.fetchFailedDevices(mainCampaign.id, request.failureCode).flatMap {
+      case deviceIds if deviceIds.isEmpty =>
+        Future.failed(MissingFailedDevices(request.failureCode))
+
+      case deviceIds =>
+        // TODO OTA-2377 we just need to create this dummy group until we get totally rid of the groups.
+        val retryGroup = GroupId(UUID.fromString("0000-00-00-00-000000"))
+        val retryCampaign = CreateRetryCampaign(mainCampaign, request.failureCode).mkCampaign
+        campaigns
+          .create(retryCampaign, Set(retryGroup), deviceIds, Nil)
```

Comment:
`Set.empty` could be used here. This way we don't need stub `retryGroup` anymore.

Reply:
Indeed. Updated.